namecoin/src/main/java/org/libdohj/names/NameLookupLatestLevelDBTransactionCache.java

```diff
@@ -0,0 +1,410 @@
+        int safetyCount;
+        for (safetyCount = 0; safetyCount < 12; safetyCount++) {
+            result = result.getPrev(store);
+        }
+        
+        return result;
+    }
+    
+    protected synchronized void putBlockChain(StoredBlock block) throws Exception {
+        
+        // TODO: use BIP 113 timestamps
+        if ( (new Date().getTime() / 1000 ) - block.getHeader().getTimeSeconds() > 366 * 24 * 60 * 60) {
+            log.debug("NameDB halting walkbalk due to timestamp expiration, height " + block.getHeight());
+            return;
+        }
+        
+        if (block.getHeight() > getChainHead() + 1) {
+            putBlockChain(block.getPrev(store));
+        }
+        
+        putBlock(block);
+    }
+    
+    // TODO: try a different peer if downloading a block fails, otherwise we're likely to stall the syncup
+    protected synchronized void putBlock(StoredBlock block) throws Exception {
+        
+        Sha256Hash blockHash = block.getHeader().getHash();
+        
+        // We might not have the block's transactions already; if we don't, we have to download the block again.
+        // This should be very rare; I'm not actually certain what circumstances would trigger it.
+        // (I guess it would happen if block.transactions is null?)
+        if (! pendingBlockTransactions.containsKey(block.getHeader().getHash())) {
+            log.warn("Transactions missing from block " + blockHash + "; re-downloading block...");
+            
+            Block nameFullBlock = peerGroup.getDownloadPeer().getBlock(blockHash).get();
+            
+            // The full block hasn't been verified in any way!
+            // So let's do that now.
+            
+            if (! nameFullBlock.getHash().equals(blockHash)) {
+                throw new Exception("Block hash mismatch!");
+            }
+            
+            // Now we know that the received block actually does match the hash that we requested.
+            // However, that doesn't mean that the block's contents are valid.
+            
+            final EnumSet<Block.VerifyFlag> flags = EnumSet.noneOf(Block.VerifyFlag.class);
+            nameFullBlock.verify(-1, flags);
+            
+            // Now we know that the block is internally valid (including the merkle root).
+            // We haven't verified signature validity, but our threat model is SPV.
+            
+            for (Transaction tx : nameFullBlock.getTransactions()) {
+                for (TransactionOutput output : tx.getOutputs()) {
+                    try {
+                        Script scriptPubKey = output.getScriptPubKey();
+                        NameScript ns = new NameScript(scriptPubKey);
+                        if(ns.isNameOp() && ns.isAnyUpdate() ) {
+                            pendingBlockTransactions.put(block.getHeader().getHash(), tx);
+                        }
+                    } catch (ScriptException e) {
+                        continue;
```

Comment:
It would be good to have a note for reference on why we're accepting transactions with invalid scripts.